File: app/src/types/information.ts
import type { Message } from "@/types/message";

// バックエンド domain.InformationAccessType に対応する。
export type InformationAccessType = "public" | "restricted";

// バックエンド domain.InformationResponsePolicy に対応する。
export type InformationResponsePolicy = "anonymous" | "authenticated";

// バックエンド usecase.StructuredOption に対応する。
export interface StructuredOption {
  value: string;
  sort_order: number;
}

// バックエンド usecase.StructuredSource に対応する。
export interface StructuredSource {
  type: string;
  key: string;
  value: string;
  status: string;
  question?: string;
  interaction_type?: string;
  options?: StructuredOption[];
}

// バックエンド usecase.StructuredInformation に対応する。
export interface StructuredInformation {
  title: string;
  sources: StructuredSource[];
}

// POST /informations/messages のリクエストボディ。
export interface ProcessInformationRequest {
  access_type: InformationAccessType;
  response_policy: InformationResponsePolicy;
  messages: Message[];
  user_input: string;
}

// POST /informations/messages のレスポンス。
// confirmedがtrueの場合のみinformation_idが設定される。
export interface ProcessInformationResponse {
  messages: Message[];
  structured: StructuredInformation | null;
  confirmed: boolean;
  information_id?: string;
}

// GET /informations/:id/recipients のレスポンス。
export interface ListRecipientsResponse {
  user_ids: string[];
}

// バックエンド usecase.OptionView に対応する。
export interface OptionView {
  id: string;
  value: string;
  sort_order: number;
}

// バックエンド usecase.SourceView に対応する。
// GET /informations/:id/sources のレスポンス項目。
// SubmitResponse(POST /informations/:id/responses)に必要なsource_id・
// interaction_type・option_idを受信者に提示するために使う。
export interface SourceView {
  id: string;
  type: string;
  key: string;
  value: string;
  interaction_type?: "radio" | "check" | "text";
  options?: OptionView[];
}

// バックエンド usecase.InformationSummaryView に対応する。
// GET /informations のレスポンス項目。
export interface InformationSummaryView {
  id: string;
  title: string;
  access_type: InformationAccessType;
  response_policy: InformationResponsePolicy;
  created_at: string;
}

File: api/internal/infrastructure/db/model/information.go
package model

import (
	"time"
	"uuid"
)

// InformationAccessType はinformationsテーブルのaccess_typeカラムの値を表す。
type InformationAccessType string

const (
	InformationAccessTypePublic     InformationAccessType = "public"
	InformationAccessTypeRestricted InformationAccessType = "restricted"
)

// InformationResponsePolicy はinformationsテーブルのresponse_policyカラムの値を表す。
type InformationResponsePolicy string

const (
	InformationResponsePolicyAnonymous     InformationResponsePolicy = "anonymous"
	InformationResponsePolicyAuthenticated InformationResponsePolicy = "authenticated"
)

// Information はinformationsテーブルの行を表す。
// 発信するユーザーが作成する最上位のレコードである。
type Information struct {
	ID              uuid.UUID
	AccountID       uuid.UUID
	CreatedByUserID uuid.UUID
	Title           string
	AccessType      InformationAccessType
	ResponsePolicy  InformationResponsePolicy
	CreatedAt       time.Time
	UpdatedAt       time.Time
}

// Recipient はrecipientsテーブルの行を表す。
// restrictedなInformationを閲覧できるUserを1件保持する。
type Recipient struct {
	ID            uuid.UUID
	InformationID uuid.UUID
	UserID        uuid.UUID
	CreatedAt     time.Time
}

// SourceType はCLAUDE.mdで定義されたSource of Truthの分類を表す。
type SourceType string

const (
	SourceTypeIdentity    SourceType = "identity"
	SourceTypeFact        SourceType = "fact"
	SourceTypeSchedule    SourceType = "schedule"
	SourceTypeCondition   SourceType = "condition"
	SourceTypeInteraction SourceType = "interaction"
)

// SourceStatus はSourceの値がどの程度確定しているかを表す。
type SourceStatus string

const (
	SourceStatusConfirmed SourceStatus = "confirmed"
	SourceStatusUndecided SourceStatus = "undecided"
	SourceStatusUnknown   SourceStatus = "unknown"
)

// SourceInteractionType は受信者がSourceに対してどう応答すべきかを表す。
// 受信者の応答を必要としないSourceの場合はnilになる。
type SourceInteractionType string

const (
	SourceInteractionTypeRadio SourceInteractionType = "radio"
	SourceInteractionTypeCheck SourceInteractionType = "check"
	SourceInteractionTypeText  SourceInteractionType = "text"
)

// Source はsourcesテーブルの行を表す。
// Informationに属し、構造化された情報を1件保持する。
type Source struct {
	ID              uuid.UUID
	InformationID   uuid.UUID
	Type            SourceType
	Key             string
	Value           string
	Status          SourceStatus
	InteractionType *SourceInteractionType
	CreatedAt       time.Time
	UpdatedAt       time.Time
}

// Option はoptionsテーブルの行を表す。
// Sourceに属し、選択肢を1件保持する。
type Option struct {
	ID        uuid.UUID
	SourceID  uuid.UUID
	Value     string
	SortOrder int
	CreatedAt time.Time
	UpdatedAt time.Time
}

// Response はresponsesテーブルの行を表す。
// Informationに対するUserの返答1件を保持する(配下のResponseItemが個々の回答を持つ)。
// UserIDはnullable(匿名回答)。
type Response struct {
	ID            uuid.UUID
	InformationID uuid.UUID
	UserID        *uuid.UUID
	CreatedAt     time.Time
	UpdatedAt     time.Time
}

// ResponseItem はresponse_itemsテーブルの行を表す。
// Responseに属し、1つのSourceに対する回答を1件保持する。
// OptionID・Valueはともにnullableで、どちらか一方のみが設定される
// (check・radioはOptionID、textはValue)。
type ResponseItem struct {
	ID         uuid.UUID
	ResponseID uuid.UUID
	SourceID   uuid.UUID
	OptionID   *uuid.UUID
	Value      *string
	CreatedAt  time.Time
}
